sympy/core/add.py

```diff
@@ -828,7 +828,27 @@ def _eval_as_leading_term(self, x):
 
         infinite = [t for t in expr.args if t.is_infinite]
 
-        expr = expr.func(*[t.as_leading_term(x) for t in expr.args]).removeO()
+        new_expr = expr.func(*[t.as_leading_term(x) for t in expr.args]).removeO()
+
+        if new_expr and (new_expr is not S.NaN):
+            #to check if the leading term cancel each other out
+            if new_expr.is_Add:
+                final_leading_term = [t for t in new_expr.args]
+            else:
+                final_leading_term = [new_expr]
+
+            if len(expr.args) != len(final_leading_term):
+                leading_term = [(t, t.as_leading_term(x)) for t in expr.args]
```

Comment:
This looks like it could use some cleaning up. You are recomputing the `leading_term` for `expr.args`...but you already did that at line 831.